PYME/cluster/rules.py

```diff
@@ -0,0 +1,627 @@
+        time.sleep(1.5)
+    
+        while (self.doPoll == True):
+            try:
+                rel_start, rel_end = self.get_new_tasks()
+                self._release_tasks(rel_start, rel_end)
+            except NoNewTasks:
+                pass
+        
+            if self.complete:
+                logging.debug('input data complete and all tasks pushed, marking rule as complete')
+                self._mark_complete()
+                logging.debug('ending polling loop.')
+                self.doPoll = False
+            else:
+                time.sleep(1)
+
+    def cleanup(self):
+        self.doPoll = False
+
+
+class RecipeRule(Rule):
+    def __init__(self, recipe=None, recipeURI=None, output_dir=None, **kwargs):
+        """
+        Create a recipe rule
+
+        Parameters
+        ----------
+        recipe : str or recipes.modules.ModuleCollection
+            The recipe as YAML text or as a recipe instance (alternatively provide recipeURL)
+        recipeURI : str
+            A cluster URI for the recipe text (if `recipe` not provided directly)
+        output_dir : str
+            The directory to put the recipe output TODO: should this be templated based on context?
+        kwargs : any additional args to get passed to base class - e.g. on_completion
+
+        TODO - support for templated recipes? Subclass?
+        """
+        
+        if recipe:
+            if isinstance(recipe, six.string_types):
+                self.recipe_text = recipe
+            else:
+                self.recipe_text = recipe.toYAML()
+            
+            self.recipeURI = None
+        else:
+            if recipeURI is None:
+                raise ValueError('recipeURI must be defined if no recipe given')
+            else:
+                self.recipeURI = recipeURI
+        
+        self.output_dir = output_dir
+
+        Rule.__init__(self, **kwargs)
+    
+    def _populate_rule(self, context, on_completion=None):
+        # over-ride here because we need to add input info
+        
+        # try (in order of increasing precedence):
+        #
+        # - an `input_templates` variable in the context
+        # - hard coded `inputs` in the context
+        inputs = {} # = context.get('rule_outputs', {})
+        input_templates = context.get('input_templates', None)
+        if input_templates is not None:
+            inputs.update({k: v.format(context) for k, v in input_templates.items()})
+        
+        inputs.update(context.get('inputs', {}))
+
+        if not inputs:
+            if not context.get('rule_outputs', False):
+                raise RuntimeError('No inputs found, one of "inputs", "input_templates", or "rule_outputs" must be present in context')
+            inputs_by_task = None
+        else:
+            input_names = inputs.keys()
+            inputs = {k: inputs[k] if isinstance(inputs[k], list) else clusterIO.cglob(inputs[k], include_scheme=True) for k
```

Comment:
annoying to have to look through multiple functions to find a required init parameter of recipe rule due to the kwarg passing. We should at least document it in the __init__ of recipe rule. `include_scheme` isn't a kwarg of clusterIO.cglob, so at the moment you are hosed if inputs[k] isn't a list.

Reply:
working on a PR